[Task:] Update presta/sitemap-bundle to version 3

kingjia90 opened the pull request:
## Changes in this pull request  
Resolves #11661

in 3.0, only the folder structure has changed, the rest seems working
https://github.com/prestaconcept/PrestaSitemapBundle/releases/tag/v3.0.0

please see upgrade notes or Demo and related [PR](https://github.com/pimcore/demo/pull/313/files)

kingjia90 committed:
fix BC config change and some broken/outdated urls

folder structure changed: https://github.com/prestaconcept/PrestaSitemapBundle/releases/tag/v3.0.0

kingjia90 committed:
composer change for sitemap bundle

jdreesen commented:
This needs an entry in the upgrade-docs, no?

jdreesen commented:
I'm just wondering if it would be possible to support both, `v2` and `v3`, leaving the decision when to upgrade to the users. It wouldn't be a BC break then.

kingjia90 commented:
> I'm just wondering if it would be possible to support both, `v2` and `v3`, leaving the decision when to upgrade to the users. It wouldn't be a BC break then.

it has some problems with PHP8.1, more details here https://github.com/pimcore/pimcore/issues/11800 and it's required if we want to update to symfony 6

kingjia90 committed:
changed sitemap bundle to the "right" version

value internally defined in this https://github.com/pimcore/product-management/issues/2

jdreesen commented:
But do you want to force-upgrade to Symfony 6 or do you just want to *allow* Symfony 6? If it's the latter, it would be fine to just allow `presta/sitemap-bundle v3`, too, and anyone who wants to use PHP 8.1 or Symfony 6 can upgrade to `presta/sitemap-bundle v3` themselves.

NiklasBr commented on doc/Development_Documentation/18_Tools_and_Features/39_Sitemaps.md:
````diff
@@ -11,12 +11,12 @@ an event handler and start to add entries.
 ```yaml
 PrestaSitemapBundle:
-    resource: "@PrestaSitemapBundle/Resources/config/routing.yaml"
+    resource: "@PrestaSitemapBundle/config/routing.yaml"
````
Also mentioned in https://github.com/pimcore/pimcore/blob/10.x/doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md?

fashxp commented:
> But do you want to force-upgrade to Symfony 6 or do you just want to _allow_ Symfony 6?

Pimcore 11 will most likely require Symfony 6

jdreesen commented:
> Pimcore 11 will most likely require Pimcore 6

That's fine, as it is a new major version which allows BC breaks. But this PR is currently assigned to the `10.5` milestone, so a graceful update should be possible.

kingjia90 committed:
added sitemap upgrade notes

fashxp commented:
if BC break is needed, it will not go into 10.5 :-) 
we'll have a look...

jdreesen commented:
If the only BC break is the new folder structure, it should be fine to support both and let the users update as desired. If this is postponed to Pimcore v11, users will not be able to fix the deprecation notices it throws on PHP 8.

kingjia90 committed:
changed all references to 11.x

brusch commented:
@jdreesen but this would still need manual work before upgrading, because by using `"presta/sitemap-bundle": "^2.1 || ^3.0"` Composer would automatically install the latest release of v3.

jdreesen commented:
That's true (if the user doesn't have a `"presta/sitemap-bundle": "^2.1"` in their `composer.json`, which they probably haven't).

brusch commented:
@kingjia90 a possible solution I can think of is to override the config using this function: 
https://github.com/pimcore/pimcore/blob/6706dba38446ab23ca2ea1a704cae261eac45fbf/bundles/CoreBundle/DependencyInjection/PimcoreCoreExtension.php#L448

We did this in the past and there's still an exampe in the code how to do that. 
So what I'm thinking of is to check there whether we are on v3 and if the config path is still `@PrestaSitemapBundle/Resources/config/routing.yml` and if so, we just rewrite the config to the new path. 

WDYT? Should we try that approach? Any concerns?

kingjia90 commented:
@brusch that would be a good solution, i'll give it a try and update you

brusch commented:
@kingjia90 btw. even if that works, we should imho still allow v2, by using `"presta/sitemap-bundle": "^2.1 || ^3.0"`

brusch commented:
Unfortunately my approach described above using `prepend()` method, doesn't work, as `resource` loading get's evaluated right on loading the config and throws an exception if the linked resource doesn't exist.

So I've another approach using a custom route loader: 
https://symfony.com/doc/current/routing/custom_route_loader.html#creating-a-custom-loader

We could implement another custom route loader, that has the following in the `supports()` method: 
```php
public function supports($resource, string $type = null)
{
    if ($resource === '@PrestaSitemapBundle/Resources/config/routing.yml') {
        try {
            $loader->getLocator()->locate('@PrestaSitemapBundle/config/routing.yml');
            return true;
        } catch (FileLocatorFileNotFoundException $e) {
            return false;
        }
    }
}
```

And in the `load()` method it just rewrites the resource path and then delegates to the existing `YamlFileLoader`. 

That should actually work 😊

kingjia90 commented:
@brusch I've give it a try, unfortunately it doesn't "detect" the Resource and breaks
![image](https://user-images.githubusercontent.com/6014195/167367191-bd7bd0d9-9e49-4aef-add8-f6d6c2f71f99.png)

but it seems could work if i manually add `type: annotation` below 
```
presta_sitemap:
    resource: "@PrestaSitemapBundle/Resources/config/routing.yml"
```
any tips?

brusch committed:
Added BC layer for old route path, support both v2 and v3, changed upgrade docs

brusch commented:
@kingjia90 please see ecedde6af003eacb42a67ee9f8d4503aef503177 - I've implemented a different approach, not super nice, but  works 😊
Can u plz double check if that is a proper bc layer? Thanks!

kingjia90 committed:
reverting doc reference to 10.x

kingjia90 committed:
small change to 3.2 and in the docs

kingjia90 commented:
The BC Layer works fine.
Changed version to [3.2](https://github.com/prestaconcept/PrestaSitemapBundle/releases/tag/v3.2.0)  as it has a better support for Symfony 6.
